src/downloader.rs

```diff
@@ -7,18 +8,62 @@ use crate::warn;
 pub struct Downloader {
     client: reqwest::blocking::Client,
     tries: usize,
+    auth_map: HashMap<String, (String, Option<String>)>,
+}
+
+/// Parse HTTP authentication credentials from string iterable
+fn parse_auth(auth: &[String], origin: &Url) -> Result<(String, Option<String>, String), String> {
+    // Convert any empty strings to None
+    let auth: Vec<Option<String>> = auth
+        .iter()
+        .map(|s| match s.as_ref() {
+            "" => None,
+            s => Some(s.to_string()),
+        })
+        .collect();
+
+    // Match on auth values and origin host, defaulting to the origin host if host not provided
+    match (auth.as_slice(), origin.host_str()) {
+        ([Some(username)], Some(origin_host)) => {
+            Ok((username.to_string(), None, origin_host.to_string()))
+        }
+        ([Some(username), password], Some(origin_host)) => Ok((
+            username.to_string(),
+            password.clone(),
+            origin_host.to_string(),
+        )),
+        ([Some(username), password, None, ..], Some(origin_host)) => Ok((
+            username.to_string(),
+            password.clone(),
+            origin_host.to_string(),
+        )),
+        ([Some(username), password, Some(host), ..], _) => {
+            Ok((username.to_string(), password.clone(), host.to_string()))
+        }
+        _ => Err("Invalid arguments supplied to auth".to_string()),
+    }
 }
 
 impl Downloader {
     /// Create a new Downloader
-    pub fn new(tries: usize, user_agent: &str) -> Downloader {
+    pub fn new(tries: usize, user_agent: &str, auth: &[String], origin: &Url) -> Downloader {
+        // Create a mapping of hosts to username, password tuples for authentication
+        let mut auth_map = HashMap::new();
+        // Iterate over the auth string in chunks of 3 items each for (username, password, host)
+        for auth_chunk in auth.chunks(3) {
```

Comment:
How about using a `const AUTH_CHUNK_SIZE: usize = 3` ? It would remove the need for a magic value and also maybe make it more future proof and easier to change in case of additions to that PR

```suggestion
        for auth_chunk in auth.chunks(AUTH_CHUNK_SIZE) {
```